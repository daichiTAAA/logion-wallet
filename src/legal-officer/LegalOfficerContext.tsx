import React, { useContext, useEffect, useReducer, Reducer } from 'react';
import { AxiosInstance } from 'axios';
import { VaultTransferRequest } from '@logion/client';

import { ProtectionRequest, } from '../common/types/ModelTypes';
import { fetchProtectionRequests, } from '../common/Model';
import { useCommonContext } from '../common/CommonContext';
import { LIGHT_MODE } from './Types';
import { AxiosFactory } from '../common/api';
import { useLogionChain } from '../logion-chain';
import { VaultApi } from '../vault/VaultApi';

export interface LegalOfficerContext {
    refreshRequests: ((clearBeforeRefresh: boolean) => void) | null,
    pendingProtectionRequests: ProtectionRequest[] | null,
    activatedProtectionRequests: ProtectionRequest[] | null,
    protectionRequestsHistory: ProtectionRequest[] | null,
    pendingRecoveryRequests: ProtectionRequest[] | null,
    recoveryRequestsHistory: ProtectionRequest[] | null,
    axios?: AxiosInstance;
    pendingVaultTransferRequests?: VaultTransferRequest[];
    vaultTransferRequestsHistory?: VaultTransferRequest[];
}

interface FullLegalOfficerContext extends LegalOfficerContext {
    dataAddress: string | null;
    fetchForAddress: string | null;
}

function initialContextValue(): FullLegalOfficerContext {
    return {
        dataAddress: null,
        fetchForAddress: null,
        refreshRequests: null,
        pendingProtectionRequests: null,
        activatedProtectionRequests: null,
        protectionRequestsHistory: null,
        pendingRecoveryRequests: null,
        recoveryRequestsHistory: null,
    };
}

const LegalOfficerContextObject: React.Context<FullLegalOfficerContext> = React.createContext(initialContextValue());

type ActionType = 'FETCH_IN_PROGRESS'
    | 'SET_DATA'
    | 'SET_CURRENT_USER'
;

interface Action {
    type: ActionType;
    dataAddress?: string;
    pendingProtectionRequests?: ProtectionRequest[];
    protectionRequestsHistory?: ProtectionRequest[];
    activatedProtectionRequests?: ProtectionRequest[];
    pendingRecoveryRequests?: ProtectionRequest[],
    recoveryRequestsHistory?: ProtectionRequest[],
    refreshRequests?: (clearBeforeRefresh: boolean) => void;
    clearBeforeRefresh?: boolean;
    axios?: AxiosInstance;
    currentAccount?: string;
    rejectRequest?: ((requestId: string, reason: string) => Promise<void>) | null;
    pendingVaultTransferRequests?: VaultTransferRequest[];
    vaultTransferRequestsHistory?: VaultTransferRequest[];
}

const reducer: Reducer<FullLegalOfficerContext, Action> = (state: FullLegalOfficerContext, action: Action): FullLegalOfficerContext => {
    switch (action.type) {
        case 'FETCH_IN_PROGRESS':
            if(action.clearBeforeRefresh!) {
                return {
                    ...state,
                    fetchForAddress: action.dataAddress!,
                    pendingProtectionRequests: null,
                    protectionRequestsHistory: null,
                };
            } else {
                return {
                    ...state,
                    fetchForAddress: action.dataAddress!,
                };
            }
        case 'SET_DATA':
            if(action.dataAddress === state.fetchForAddress) {
                return {
                    ...state,
                    fetchForAddress: null,
                    dataAddress: action.dataAddress!,
                    pendingProtectionRequests: action.pendingProtectionRequests!,
                    protectionRequestsHistory: action.protectionRequestsHistory!,
                    activatedProtectionRequests: action.activatedProtectionRequests!,
                    pendingRecoveryRequests: action.pendingRecoveryRequests!,
                    recoveryRequestsHistory: action.recoveryRequestsHistory!,
                    pendingVaultTransferRequests: action.pendingVaultTransferRequests!,
                    vaultTransferRequestsHistory: action.vaultTransferRequestsHistory!,
                };
            } else {
                return state;
            }
        case "SET_CURRENT_USER": {
            return {
                ...state,
                refreshRequests: action.refreshRequests!,
                axios: action.axios!,
            };
        }
        default:
            /* istanbul ignore next */
            throw new Error(`Unknown type: ${action.type}`);
    }
}

export interface Props {
    children: JSX.Element | JSX.Element[] | null
}

export function LegalOfficerContextProvider(props: Props) {
    const { accounts, axiosFactory, isCurrentAuthenticated } = useLogionChain();
    const { colorTheme, setColorTheme } = useCommonContext();
    const [ contextValue, dispatch ] = useReducer(reducer, initialContextValue());

    useEffect(() => {
        if(colorTheme !== LIGHT_MODE && setColorTheme !== null) {
            setColorTheme(LIGHT_MODE);
        }
    }, [ colorTheme, setColorTheme ]);

    useEffect(() => {
        if(accounts !== null
                && axiosFactory !== undefined
                && accounts.current !== undefined
                && contextValue.dataAddress !== accounts.current.address
                && contextValue.fetchForAddress !== accounts.current.address
                && isCurrentAuthenticated()) {
            const currentAccount = accounts!.current!.address;
            const refreshRequests = (clear: boolean) => refreshRequestsFunction(clear, currentAccount, dispatch, axiosFactory);
            dispatch({
                type: 'SET_CURRENT_USER',
               axios: axiosFactory(currentAccount),
               currentAccount,
               refreshRequests,
            });
            refreshRequests(true);
        }
    }, [ contextValue, axiosFactory, accounts, isCurrentAuthenticated ]);

    return (
        <LegalOfficerContextObject.Provider value={contextValue}>
            {props.children}
        </LegalOfficerContextObject.Provider>
    );
}

export function useLegalOfficerContext(): LegalOfficerContext {
    return { ...useContext(LegalOfficerContextObject) };
}

function refreshRequestsFunction(clearBeforeRefresh: boolean, currentAddress: string, dispatch: React.Dispatch<Action>, axiosFactory: AxiosFactory) {
    dispatch({
        type: "FETCH_IN_PROGRESS",
        dataAddress: currentAddress,
        clearBeforeRefresh,
    });

    (async function() {
        const axios = axiosFactory(currentAddress);
        const pendingProtectionRequests = await fetchProtectionRequests(axios, {
            statuses: ["PENDING"],
            kind: 'PROTECTION_ONLY',
        });
        const activatedProtectionRequests = await fetchProtectionRequests(axios, {
            statuses: ["ACTIVATED"],
            kind: 'ANY',
        });
        const protectionRequestsHistory = await fetchProtectionRequests(axios, {
            statuses: ["ACCEPTED", "REJECTED", "ACTIVATED"],
            kind: 'PROTECTION_ONLY',
        });

        const pendingRecoveryRequests = await fetchProtectionRequests(axios, {
            statuses: ["PENDING"],
            kind: 'RECOVERY',
        });
        const recoveryRequestsHistory = await fetchProtectionRequests(axios, {
            statuses: ["ACCEPTED", "REJECTED", "ACTIVATED"],
            kind: 'RECOVERY',
        });

        const vaultSpecificationFragment = {
            statuses: []
        }

        const vaultTransferRequestsResult = await new VaultApi(axios, currentAddress).getVaultTransferRequests({
            ...vaultSpecificationFragment,
            statuses: [ "PENDING" ]
        });
        const pendingVaultTransferRequests = vaultTransferRequestsResult.sort((a, b) => b.createdOn.localeCompare(a.createdOn));

        const vaultTransferRequestsHistoryResult = await new VaultApi(axios, currentAddress).getVaultTransferRequests({
            ...vaultSpecificationFragment,
            statuses: [ "CANCELLED", "REJECTED_CANCELLED", "REJECTED", "ACCEPTED" ]
        });
        const vaultTransferRequestsHistory = vaultTransferRequestsHistoryResult.sort((a, b) => b.createdOn.localeCompare(a.createdOn));

        dispatch({
            type: "SET_DATA",
            pendingProtectionRequests,
            activatedProtectionRequests,
            protectionRequestsHistory,
            dataAddress: currentAddress,
            pendingRecoveryRequests,
            recoveryRequestsHistory,
            pendingVaultTransferRequests,
            vaultTransferRequestsHistory,
        });
    })();
}