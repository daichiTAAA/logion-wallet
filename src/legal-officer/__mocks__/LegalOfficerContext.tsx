export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = () => {

};

export let pendingProtectionRequests: any[] | null = null;

export let acceptedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export function useLegalOfficerContext() {
    return {
        pendingTokenizationRequests,
        rejectRequest,
        rejectedTokenizationRequests,
        acceptedTokenizationRequests,
        refreshRequests,
        pendingProtectionRequests,
        acceptedProtectionRequests,
        protectionRequestsHistory,
    };
}

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setAcceptedProtectionRequests(requests: any[]) {
    acceptedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}
