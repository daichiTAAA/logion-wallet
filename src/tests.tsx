import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
