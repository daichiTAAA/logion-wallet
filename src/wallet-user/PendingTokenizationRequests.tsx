import React from 'react';

import Table, { Cell, EmptyTableMessage, DateCell } from '../common/Table';

import { useUserContext } from './UserContext';

export default function PendingTokenizationRequests() {
    const { pendingTokenizationRequests } = useUserContext();

    if(pendingTokenizationRequests === null) {
        return null;
    }

    return (
        <>
            <Table
                columns={[
                    {
                        header: "Legal officer",
                        render: request => <Cell content={ request.legalOfficerAddress } />,
                        align: 'left',
                    },
                    {
                        header: "Token name",
                        render: request => <Cell content={ request.requestedTokenName } />,
                        align: 'left',
                    },
                    {
                        header: "Bars",
                        render: request => <Cell content={ request.bars } />,
                        align: 'right',
                        width: '200px',
                    },
                    {
                        header: "Creation date",
                        render: request => <DateCell dateTime={ request.createdOn || null } />,
                        width: '200px',
                    }
                ]}
                data={ pendingTokenizationRequests }
                renderEmpty={ () => <EmptyTableMessage>No pending tokenization request</EmptyTableMessage> }
            />
        </>
    );
}
