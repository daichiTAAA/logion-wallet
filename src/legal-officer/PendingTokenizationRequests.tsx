import React, { useState } from 'react';

import Table from 'react-bootstrap/Table';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import Form from 'react-bootstrap/Form';

import { useLegalOfficerContext } from './LegalOfficerContext';

export default function PendingTokenizationRequests() {
    const { pendingTokenizationRequests, rejectRequest } = useLegalOfficerContext();
    const [ requestToReject, setRequestToReject ] = useState<string | null>(null);
    const [ reason, setReason ] = useState<string>("");

    if (pendingTokenizationRequests === null) {
        return null;
    }

    const handleClose = () => setRequestToReject(null);
    const rejectAndCloseModal = async () => {
        await rejectRequest!(requestToReject!, reason);
        setRequestToReject(null);
    };

    return (
        <>
            <h1>Pending Tokenization Requests</h1>
            <Table striped bordered>
                <thead>
                    <tr>
                        <th>Requester</th>
                        <th>Token</th>
                        <th>Bars</th>
                        <th>Actions</th>
                    </tr>
                </thead>
                <tbody>
                    {
                        pendingTokenizationRequests.map(request => (
                            <tr key={request.id}>
                                <td>{request.requesterAddress}</td>
                                <td>{request.requestedTokenName}</td>
                                <td>{request.bars}</td>
                                <td>
                                    <ButtonGroup aria-label="actions">
                                        <Button
                                            variant="danger"
                                            onClick={() => setRequestToReject(request.id)}
                                            data-testid={`reject-${request.id}`}
                                        >
                                            X
                                        </Button>
                                    </ButtonGroup>
                                </td>
                            </tr>
                        ))
                    }
                </tbody>
            </Table>

            <Modal show={requestToReject !== null} onHide={handleClose}>
                <Modal.Header closeButton>
                    <Modal.Title>Rejecting {requestToReject}</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <Form.Group controlId="exampleForm.ControlTextarea1">
                        <Form.Label>Reason</Form.Label>
                        <Form.Control as="textarea" rows={3} onChange={e => setReason(e.target.value)} value={reason} data-testid="reason" />
                    </Form.Group>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="secondary" onClick={handleClose}>
                        Cancel
                    </Button>
                    <Button variant="primary" onClick={rejectAndCloseModal} data-testid={`confirm-reject-${requestToReject}`}>
                        Reject
                    </Button>
                </Modal.Footer>
            </Modal>
        </>
    );
}
