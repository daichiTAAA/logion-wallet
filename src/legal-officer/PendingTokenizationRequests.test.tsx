jest.mock('./LegalOfficerContext');

import { shallowRender } from '../tests';
import PendingTokenizationRequests from './PendingTokenizationRequests';
import { setPendingRequests, setRejectRequest } from './LegalOfficerContext';
import { render, act, fireEvent } from '@testing-library/react';

test("Renders null with no data", () => {
    const tree = shallowRender(<PendingTokenizationRequests />);
    expect(tree).toMatchSnapshot();
});

test("Renders pending requests", () => {
    setPendingRequests([
        {
            id: "1",
            legalOfficerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
            requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
            requestedTokenName: "TOKEN1",
            bars: 1,
            status: "PENDING"
        }
    ]);
    const tree = shallowRender(<PendingTokenizationRequests />);
    expect(tree).toMatchSnapshot();
});

test("Click on reject and confirm rejects request", () => {
    setPendingRequests([
        {
            id: "1",
            legalOfficerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
            requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
            requestedTokenName: "TOKEN1",
            bars: 1,
            status: "PENDING"
        }
    ]);
    const rejectCallback = jest.fn();
    setRejectRequest(rejectCallback);

    const tree = render(<PendingTokenizationRequests />);
    const rejectButton = tree.getByTestId("reject-1");
    act(() => { fireEvent.click(rejectButton) });

    const reasonText = "Because";
    const reasonTextArea = tree.getByTestId("reason");
    act(() => { fireEvent.change(reasonTextArea, {target: {value: reasonText}}) });

    const confirmButton = tree.getByTestId("confirm-reject-1");
    act(() => { fireEvent.click(confirmButton) });

    expect(rejectCallback).toBeCalledWith("1", reasonText);
});
