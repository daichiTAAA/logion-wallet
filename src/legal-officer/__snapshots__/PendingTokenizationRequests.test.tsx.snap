// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders null with no data 1`] = `null`;

exports[`Renders pending requests 1`] = `
<React.Fragment>
  <h1>
    Pending Tokenization Requests
  </h1>
  <ForwardRef
    bordered={true}
    striped={true}
  >
    <thead>
      <tr>
        <th>
          Requester
        </th>
        <th>
          Token
        </th>
        <th>
          Bars
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW
        </td>
        <td>
          TOKEN1
        </td>
        <td>
          1
        </td>
        <td>
          <ButtonGroup
            aria-label="actions"
            role="group"
            toggle={false}
            vertical={false}
          >
            <Button
              active={false}
              data-testid="reject-1"
              disabled={false}
              onClick={[Function]}
              variant="danger"
            >
              X
            </Button>
          </ButtonGroup>
        </td>
      </tr>
    </tbody>
  </ForwardRef>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    onHide={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        Rejecting 
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <FormGroup
        controlId="exampleForm.ControlTextarea1"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          Reason
        </FormLabel>
        <FormControl
          as="textarea"
          data-testid="reason"
          onChange={[Function]}
          rows={3}
          value=""
        />
      </FormGroup>
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="secondary"
      >
        Cancel
      </Button>
      <Button
        active={false}
        data-testid="confirm-reject-null"
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Reject
      </Button>
    </ModalFooter>
  </Modal>
</React.Fragment>
`;
